subsys/bluetooth/mesh/crypto_mtls.c

```diff
@@ -357,3 +296,145 @@ __weak psa_status_t mbedtls_psa_external_get_random(mbedtls_psa_external_random_
 
 	return PSA_SUCCESS;
 }
+
+__weak psa_key_id_t bt_mesh_user_keyid_alloc(void)
+{
+	for (int i = 0; i < KEY_ID_RANGE_SIZE; i++) {
+		if (pst_key_id[i] == PSA_KEY_ID_NULL) {
+			pst_key_id[i] = PSA_KEY_ID_USER_MIN + i;
+			return pst_key_id[i];
+		}
+	}
+
+	return PSA_KEY_ID_NULL;
+}
+
+__weak int bt_mesh_user_keyid_free(psa_key_id_t key_id)
+{
+	int err = -EIO;
+
+	if (IN_RANGE(key_id, PSA_KEY_ID_USER_MIN,
+			PSA_KEY_ID_USER_MIN + KEY_ID_RANGE_SIZE - 1)) {
+		pst_key_id[key_id - PSA_KEY_ID_USER_MIN] = PSA_KEY_ID_NULL;
+		err = 0;
+	}
+
+	return err;
+}
+
+int bt_mesh_import_key(enum bt_mesh_key_type type, const bt_mesh_raw_key_t in, bt_mesh_key_t *out)
+{
+	psa_key_attributes_t key_attributes = PSA_KEY_ATTRIBUTES_INIT;
+	psa_status_t status;
+	int err = 0;
+
+	switch (type) {
+	case ECB_KEY:
+		psa_set_key_lifetime(&key_attributes, PSA_KEY_LIFETIME_VOLATILE);
+		psa_set_key_usage_flags(&key_attributes,
+			PSA_KEY_USAGE_ENCRYPT | PSA_KEY_USAGE_DECRYPT);
+		psa_set_key_algorithm(&key_attributes, PSA_ALG_ECB_NO_PADDING);
+		break;
+	case CCM_KEY:
+		psa_set_key_lifetime(&key_attributes, PSA_KEY_LIFETIME_VOLATILE);
+		psa_set_key_usage_flags(&key_attributes,
+			PSA_KEY_USAGE_ENCRYPT | PSA_KEY_USAGE_DECRYPT);
+		psa_set_key_algorithm(&key_attributes,
+			PSA_ALG_AEAD_WITH_AT_LEAST_THIS_LENGTH_TAG(PSA_ALG_CCM, 4));
+		break;
+	case CMAC_KEY:
+		psa_set_key_lifetime(&key_attributes, PSA_KEY_LIFETIME_VOLATILE);
+		psa_set_key_usage_flags(&key_attributes, PSA_KEY_USAGE_SIGN_MESSAGE);
+		psa_set_key_algorithm(&key_attributes, PSA_ALG_CMAC);
+		break;
+	case NETWORK_KEY:
+		if (IS_ENABLED(CONFIG_BT_SETTINGS)) {
+			psa_key_id_t key_id = bt_mesh_user_keyid_alloc();
```

Comment:
Good point. I don't import in `bt_mesh_subnet_set ` by design because my assumption was it has been allocated by the application in advance. Case when it is restored from cdb I missed. This is the obvious issue. If you don't have an objection I'd suggest to solve this as a separate task. It seems quite an autonomic task that could be done separately.

Reply:
To be addressed in NCSDK-17765